# merge alignment data from aligned BAM files with molecular index data from 
# unmapped BAM files (generated by picard IlluminaBasecallsToSam). They were not aligned to any genome, 
# the reads are all contail UMIs, so they can't be mapped and then called "unmapped/unaligned bams"
rule merge_bam:
    input:
        mapped = "STAR_align_multi/{sample}.bam",
        unmapped= "unmapped_bam/{sample}_unaligned.bam"
    output:
        "merged_bam/{sample}_merged.bam"
    params:
        ref_seq = genome_seq,
        mem = '10G',
        jobName = "merge_bam.{sample}"
    log: "logs/merge_bam/{sample}.log"     
    threads: 4
    shell:
    # use the parameters in RiboToolkit
        '''
        java -Xmx8g -jar ~/public/tools/picard-2.25.1-1/picard.jar MergeBamAlignment \\
        --ALIGNED {input.mapped} --UNMAPPED {input.unmapped} \\
        --REFERENCE_SEQUENCE {params.ref_seq} --OUTPUT {output} --MAX_GAPS -1 \\
        --SORT_ORDER coordinate --CREATE_INDEX false &>> {log}
        '''

rule merge_bam_index:
    input:
        "merged_bam/{sample}_merged.bam"
    output:
        "merged_bam/{sample}_merged.bam.bai"
    params: 
      mem = '20G',
      jobName = "merge_bam_index.{sample}"
    log: "logs/merge_bam_index/{sample}.log"       
    shell:
        "samtools index {input} {output}"